PERF: Optimize the performance of `TopicLink.topic_map` query

tgxworld opened the pull request:
Instead of relying on the `ILIKE` operator to filter out image links, we
can instead rely on the `TopicLink#extension` column which allows us to
more efficiently filter out image links.

This optimization mainly affects topics that are link heavy which is
common in topics with alot of replies. When profiling a production
instance for a topic with 10K replies and 2.5K `topic_links`, this
optimization reduces the query time from ~18ms to around ~4ms.
### Sample query runtime for said production instance above

![Screenshot from 2022-12-27 10-18-25](https://user-images.githubusercontent.com/4335742/209600794-126329ab-419a-4b22-b697-9aff10c76003.png)

Drenmi reviewed (approved):
![perfect-mwah](https://user-images.githubusercontent.com/5259935/209600692-009c7591-d924-4741-8560-18325b1bbd85.gif)

tgxworld commented on app/models/topic_link.rb:
```diff
@@ -58,8 +58,7 @@ def self.topic_map(guardian, topic_id)
     builder.where('ft.deleted_at IS NULL')
-    # note that ILIKE means "case insensitive LIKE"
-    builder.where("NOT(ftl.url ILIKE '%.png' OR ftl.url ILIKE '%.jpg' OR ftl.url ILIKE '%.gif')")
+    builder.where("ftl.extension NOT IN ('png','jpg','gif')")
```
Note that I'm not making any opinions about whether this extension list is complete as the extensions we use here has been added 5 years ago. 

If I were to give my opinion... I think there may be other file type extensions we may want to exclude...

tgxworld commented:
> we can instead rely on the TopicLink#extension column which allows us to
add an index for better query performance that was not possible when
using the ILIKE operator.

Looks like I made an incorrect assumption here. Looking at the query plan, it doesn't seem like PG can efficiently make use of the new index I added. 

```
->  Index Scan using index_topic_links_on_topic_id_and_extension_and_clicks on topic_links ftl  (cost=0.42..1179.47 rows=1 width=179) (actual time=0.014..0.735 rows=40 loops=1)
      Index Cond: ((topic_id = 477) AND (clicks > 0))
      Filter: ((extension)::text <> ALL ('{png,jpg,gif}'::text[]))
      Rows Removed by Filter: 790
```

The filtering still has to be done :thinking:. I guess the speed up here comes from the fact that filtering using `NOT IN` is much faster than having to filter using the `ILIKE` operator 3 times per row. 

@Drenmi I've simplified the changes in this PR and updated the PR description and commit message to reflect the new changes.

Drenmi commented:
Sounds even better, right? Faster, but without the overhead of an index. 🙂

tgxworld commented:
Thank you for reviewing @Drenmi
